whiptail.1
.TH WHIPTAIL 1 "31 January 2007" "Whiptail Version 0.52.5"
.SH NAME
whiptail \- display dialog boxes from shell scripts
.SH SYNOPSIS
.B whiptail
[
.B \-\-title
.I title
]
[
.B \-\-backtitle
.I backtitle
]
[
.B \-\-clear
]
[
.B \-\-default\-item
.I string
]
[
.B \-\-defaultno
]
[
.B \-\-fb
]
[
.B \-\-nocancel
]
[
.B \-\-yes\-button
.I text
]
[
.B \-\-no\-button
.I text
]
[
.B \-\-ok\-button
.I text
]
[
.B \-\-cancel\-button
.I text
]
[
.B \-\-noitem
[
]
.B \-\-output\-fd
.I fd
]
[
.B \-\-separate\-output
]
[
.B \-\-scrolltext
]
[
.B \-\-topleft
]
.B box-options
.SH DESCRIPTION
.B whiptail
is a program that will let you present a variety of questions or
display messages using dialog boxes from a shell script. Currently,
these types of dialog boxes are implemented: 
.LP
.BR yes/no " box," " menu" " box," " input" " box,"
.BR message " box," " text" " box," " info" " box,"
.BR checklist " box," " radiolist" " box" " gauge" " box, and"
.BR password " box."
.SH OPTIONS
.TP
.B \-\-clear
The screen will be cleared to the
.BR "screen attribute" " on exit."
This doesn't work in an xterm (and descendants) if alternate screen
switching is enabled, because in that case slang writes to (and clears)
an alternate screen.
.TP
.B \-\-defaultno
The dialog box will open with the cursor over the 
.BR No " button."
.TP
.BI \-\-default\-item " string"
Set  the default item in a menu box.
Normally the first item in the box is the default.
.TP
\fB\-\-fb\fR, \fB\-\-fullbuttons\fR
Use full buttons. (By default, 
.B whiptail
uses compact buttons). 
.TP
.B \-\-nocancel
The dialog box won't have a 
.BR Cancel " button".
.TP
.BI \-\-yes\-button " text"
Set the text of the
.BR Yes " button."
.TP
.BI \-\-no\-button " text"
Set the text of the
.BR No " button."
.TP
.BI \-\-ok\-button " text"
Set the text of the
.BR Ok " button."
.TP
.BI \-\-cancel\-button " text"
Set the text of the
.BR Cancel " button."
.TP
.B \-\-noitem
The menu, checklist and radiolist widgets will display tags only, not
the item strings. The menu widget still needs some items specified,
but checklist and radiolist expect only tag and status.
.TP
.B \-\-notags
Don't display tags in the menu, checklist and radiolist widgets.
.TP
.BI \-\-separate\-output
For checklist widgets, output result one line at a time, with no
quoting.  This facilitates parsing by another program.
.TP
.BI \-\-output\-fd  " fd" 
Direct output to the given file descriptor.  Most 
.B whiptail
scripts
write to standard error, but  error  messages  may  also  be
written there, depending on your script.
.TP
.BI \-\-title " title"
Specifies a
.I title
string to be displayed at the top of the dialog box.
.TP
.BI \-\-backtitle " backtitle"
Specifies a
.I backtitle
string to be displayed on the backdrop, at the top of the screen.
.TP
.BI \-\-scrolltext
Force the display of a vertical scrollbar.
.TP
.BI \-\-topleft
Put window in top-left corner.
.TP
\fB\-h\fR, \fB\-\-help\fR
Print a help message and exit.
.TP
\fB\-v\fR, \fB\-\-version\fR
Print version information and exit.
.TP
.B Box Options
.TP
.BI \-\-yesno " text height width"
.RB A " yes/no" " dialog box of size"
.I height
rows by
.I width
columns will be displayed. The string specified by
.I text
is displayed inside the dialog box. If this string is too long to be fit
in one line, it will be automatically divided into multiple lines at
appropriate places. The
.I text
string may also contain the sub-string
.I
"\en"
or newline characters
.I `\en'
to control line breaking explicitly.  This dialog box is useful for
asking questions that require the user to answer either yes or no.
.RB "The dialog box has a" " Yes" " button and a " No
button, in which the user can switch between by pressing the
.IR TAB " key."
.TP
.BI \-\-msgbox " text height width"
.RB A " message" " box is very similar to a" " yes/no" " box."
The only difference between a
.B message
box and a
.B yes/no
box is that a
.B message
box has only a single
.B OK
button. You can use this dialog box to display any message you like.
After reading the message, the user can press the
.I ENTER
key so that
.B whiptail
will exit and the calling shell script can continue its operation.
.TP
.BI \-\-infobox " text height width"
.RB An " info" " box is basically a" " message" " box."
However, in this case,
.B whiptail
will exit immediately after displaying the message to the user. The
screen is not cleared when
.B whiptail
exits, so that the message will remain on the screen until the calling
shell script clears it later. This is useful when you want to inform
the user that some operations are carrying on that may require some
time to finish.
.TP
.BI \-\-inputbox " text height width [init]"
.RB "An " input " box is useful when you want to ask questions that"
require the user to input a string as the answer. If init is supplied
it is used to initialize the input string.
When inputing the
string, the
.I BACKSPACE
key can be used to correct typing errors. If the input string is longer than
the width of the dialog box, the input field will be scrolled. On exit,
the input string will be printed on
.IR stderr "."
.TP
.BI \-\-passwordbox " text height width [init]"
.RB "A " password " box is similar to an input box, except the text the user"
enters is not displayed. This is useful when prompting for passwords or other
sensitive information. Be aware that if anything is passed in "init", it
will be visible in the system's process table to casual snoopers. Also, it
is very confusing to the user to provide them with a default password they
cannot see. For these reasons, using "init" is highly discouraged.
.TP
.BI \-\-textbox " file height width"
.RB A " text" " box lets you display the contents of a text file in a"
dialog box. It is like a simple text file viewer. The user can move
through the file by using the
.IR UP/DOWN ", " PGUP/PGDN
.RI and " HOME/END" " keys available on most keyboards."
If the lines are too long to be displayed in the box, the
.I LEFT/RIGHT
keys can be used to scroll the text region horizontally. For more
convenience, forward and backward searching functions are also provided.
.IP "\fB\-\-menu \fItext height width menu-height \fR[ \fItag item \fR] \fI..."
As its name suggests, a
.B menu
box is a dialog box that can be used to present a list of choices in
the form of a menu for the user to choose. Each menu entry consists of a
.IR tag " string and an " item " string. The"
.I tag
gives the entry a name to distinguish it from the other entries in the
menu. The
.I item
is a short description of the option that the entry represents. The
user can move between the menu entries by pressing the
.I UP/DOWN
keys, the first letter of the
.I tag
as a hot-key. There are
.I menu-height
entries displayed in the menu at one time, but the menu will be
scrolled if there are more entries than that. When
.B whiptail
exits, the
.I tag
of the chosen menu entry will be printed on
.IR stderr "."
.IP "\fB\-\-checklist \fItext height width list-height \fR[ \fItag item status \fR] \fI..."
.RB "A " checklist " box is similar to a " menu " box in that there are"
multiple entries presented in the form of a menu.
You can select and deselect items using the SPACE key.  
The initial on/off state of each entry is specified by
.IR status "."
On exit, a list of the
.I tag
strings of those entries that are turned on will be printed on
.IR stderr "."

.IP "\fB\-\-radiolist \fItext height width list-height \fR [ \fItag item status \fR] \fI..."
.RB "A " radiolist " box is similar to a " menu " box.  The only difference is"
that you can indicate which entry is currently selected, by setting its
.IR status " to " on "."

.IP "\fB\-\-gauge \fItext height width percent\fR"
.RB "A " gauge " box displays a meter along the bottom of the box.
The meter indicates a percentage.  New percentages are read from
standard input, one integer per line.  The meter is updated
to reflect each new percentage.  If stdin is XXX, the first following line is
a percentage and subsequent lines up to another XXX are used for a new prompt.
The gauge exits when EOF is reached on stdin.

.SH NOTES
whiptail interprets arguments starting with a dash "\-" as being arguments.
To avoid this, and start some text in, for example, a menubox item, with a 
dash, whiptail honours the getopt convention of accepting the special
argument "\-\-" which means that all following arguments with dashes are to
be treated verbatim and not parsed as options.

If the specified height, width, or menu/list-height is zero,
.BR whiptail
will increase the values as needed in order to display all information.

.SH DIAGNOSTICS
Exit status is 0 if
.BR whiptail " is exited by pressing the " Yes " or " OK
button, and 1 if the
.BR No " or " Cancel
button is pressed. Otherwise, if errors occur inside
.B whiptail
or
.B whiptail
is exited by pressing the
.I ESC
key, the exit status is -1.
.SH AUTHOR
Based on the man page for dialog(1) by:
.LP
Savio Lam (lam836@cs.cuhk.hk) - version 0.3
.LP
Stuart Herbert (S.Herbert@sheffield.ac.uk) - patch for version 0.4
.LP
Modifications for whiptail by:
.LP
Enrique Zanardi (ezanard@debian.org)
.LP
Alastair McKinstry (mckinstry@debian.org)


CHANGES
0.52.23
- fix automatic height of menu/list in whiptail (broken in 0.52.22)
- fix automatic width of whiptail --yesno box
- fix automatic width in whiptail with unicode characters
- fix automatic width with whiptail --noitem and --notags options
- fix spacing with longer tags in whiptail
- avoid overlapping backtitle in whiptail with automatic height

0.52.22
- fix crash in whiptail with new libpopt
- switch from usleep to nanosleep (Rosen Penev)
- fix libnewt.pc to enable static linking (Alexey Sheplyakov)
- fix LDFLAGS order in snack linking (Sam James)
- use CFLAGS when compiling snack
- improve configure.ac (Thomas Kuehne)
- install header and libnewt.pc with shared library (Michael Olbrich)

0.52.21
- define env NEWT_NOFLOWCTRL to disable flow control (Robert Gill)
- don't leak memory on errors in dialogboxes
- fix radio button selection check in snack
- fix plural forms in Lithuanian translation (#1568999)
- fix parallel build (Émeric Dupont)
- allow python versions to be specified with --with-python option

0.52.20
- improve handling of long strings in whiptail menu (#1353792)
- support screen resize as hotkey in form (#1432926)

0.52.19
- support --notags in whiptail checklist and radiolist (#1319794)
- don't hard-code -I/usr/include/slang in CPPFLAGS (Alex Suykov)
- use $(CC) instead of $(CPP) to generate .depend files (Samuel Martin)
- update translations from Zanata

0.52.18
- fix widget key collision in snack on 64-bit archs (#1151455)
- handle NEWT_EXIT_ERROR in snack
- fix preprocessor warnings in snack
- optimize textbox reflowing
- remove newtListitem declarations (#1074092)

0.52.17
- add python3 support (Matthias Klose) (#963839)
- implement newtComponentAddCallback() for forms as a focus-change
  callback (Dan Winship)
- add newtEntryGet/SetCursorPosition and newtFormGet/SetScrollPosition
  (Dan Winship)
- restore cursor position after changing help line (Dan Winship)
- rename snackmodule to snack (#963839)
- update CHANGES from spec changelog
- add AUTHORS and README

0.52.16
- add newtComponentGetSize and newtComponentGetPosition
  (Dan Winship) (#987596)
- modify Makefile to use SOEXT (#971168)
- free gpm socket name and unlink gpm socket on form exit
- fix memory leaks in whiptail
- add missing whiptail options to help and man page

0.52.15
- fix errors found by gcc-with-cpychecker (#800075)
- fix building with tcl8.6 (#902561)
- add fallback to python-config (#783627)
- replace tabs in snack.py (#870647)
- compile snackmodule.c with flag -fPIC (Kang Kai)
- include new translations from transifex
- allow newtWinMenu and newtWinEntries with no buttons or items
- don't draw scale when not mapped
- build with large-file support for stat64
- remove unused variables in test code
- update FSF address

0.52.14
- fix returning strings in whiptail and whiptcl (#752818)
- fix configure to work with multiple python versions (#737998)

0.52.13
- add support for changing colors in individual labels, scrollbars, entries,
  textboxes and scales, add custom colorsets 
- add support for NEWT_COLORS and NEWT_COLORS_FILE variables (#689903)
- allow resizing of form
- fix errors found by coverity
- fix va_list usage (Gwenole Beauchesne)
- fix building and installing on Mac OS X (#652479)
- check for slang.h header, support DESTDIR variable, add --without-python
  option (Otavio Salvador)
- add Persian, Low German translations
- don't hang in form when stdin disappears

0.52.12
- fix whiptail --gauge and its description in man page (#620083)
- remove space after \n in whiptail texts (#620083)
- remove NLS code from snack (#599608)
- expose more keys to python as shortcuts in dialogs (Jakob Kemi)
- release python global-thread-lock during dialog displays (Jakob Kemi)
- fix warnings in whiptcl.c and include Tcl_PkgProvide() call (Mikhail T.)
- don't NULL deref when an invalid array is specified in checkboxtree
  (Arnaldo Carvalho de Melo)

0.52.11
- fix buffer overflow in textbox when reflowing (#523955, CVE-2009-2905)
- use full textbox width when reflowing and allow minimal width 1
- fix writing lines longer than width in textbox
- don't use va_list in newtvwindow more than once (#523696)
- bind \E[Z to back-tab in built-in keymap (#468046)
- terminate string after reading file in whiptail
- add newtRadioSetCurrent function (Thomas Jarosch)
- add pkgconfig support (Thomas Jarosch)
- add Malay, Malayalam, Assamese, Gujarati, Bengali India, Kannada, Telugu
  translations
- include debian patches
  - fix crash in textbox SetText when topLines != 0
  - don't link modules with libraries already linked with libnewt
  - add Asturian and Marathi translations

0.52.10
- improve --noitem description (#456305)
- add setHeight to Textbox class
- fix fixedheight forms
- free keymap in newtFinished()
- fix memory leak in textbox
- fix valgrind error in checkboxtree
- don't crash when running empty form
- don't crash or hang when form has no focusable elements
- before checkboxtree drawing return first item in GetCurrent()
- redraw textbox in SetText()
- add setColor description to SnackScreen docstring (Greg Swift)
- make sure Widget isn't used directly (Greg Swift) (#452920)
- add Serbian translations (Miloš Komarčević)
- add Balochi translation (Mostafa Daneshvar)

0.52.9
- handle component destruction (patch by Richard W.M. Jones)
- fix newtWinEntry definition
- don't use uninitialized values in newtWinMenu
- remove workarounds for old bug in SLsmg_write_nstring
- improve SIGWINCH handling in form
- don't abort from whiptail gauge on SIGWINCH
- redisplay also last line
- update Polish translation (Piotr Drąg)

0.52.8
- enable slang utf8 mode (#425992)
- support --disable-nls option (patch by Natanael Copa)
- redraw screen when using entry in euc encodings
- fix segfault in whiptail when no entry is selected in radiolist
- add back support for list of Entries in EntryWindow prompts in snack
  (#248878)
 
0.52.7
- add support to snack for multiple selection and border in listbox

whiptail.c
#include "config.h"
#include <fcntl.h>
#include <popt.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>
#include <wchar.h>
#include <slang.h>
#include <sys/stat.h>

#include "nls.h"
#include "dialogboxes.h"
#include "newt.h"
#include "newt_pr.h"

enum { NO_ERROR = 0, WAS_ERROR = 1 };

enum mode { MODE_NONE, MODE_INFOBOX, MODE_MSGBOX, MODE_YESNO, MODE_CHECKLIST,
		MODE_INPUTBOX, MODE_RADIOLIST, MODE_MENU, MODE_GAUGE ,
		MODE_TEXTBOX, MODE_PASSWORDBOX};

#define OPT_MSGBOX 		1000
#define OPT_CHECKLIST 		1001
#define OPT_YESNO 		1002
#define OPT_INPUTBOX 		1003
#define OPT_FULLBUTTONS 	1004
#define OPT_MENU	 	1005
#define OPT_RADIOLIST	 	1006
#define OPT_GAUGE	 	1007
#define OPT_INFOBOX	 	1008
#define OPT_TEXTBOX		1009
#define OPT_PASSWORDBOX		1010

static void usage(int err) {
    newtFinished();
    fprintf (err ? stderr : stdout,
             _("Box options: \n"
	       "\t--msgbox <text> <height> <width>\n"
	       "\t--yesno  <text> <height> <width>\n"
	       "\t--infobox <text> <height> <width>\n"
	       "\t--inputbox <text> <height> <width> [init] \n"
	       "\t--passwordbox <text> <height> <width> [init] \n"
	       "\t--textbox <file> <height> <width>\n"
	       "\t--menu <text> <height> <width> <listheight> [tag item] ...\n"
	       "\t--checklist <text> <height> <width> <listheight> [tag item status]...\n"
	       "\t--radiolist <text> <height> <width> <listheight> [tag item status]...\n"
	       "\t--gauge <text> <height> <width> <percent>\n"
	       "Options: (depend on box-option)\n"
	       "\t--clear				clear screen on exit\n"
	       "\t--defaultno			default no button\n"	
	       "\t--default-item <string>		set default string\n"
	       "\t--fb, --fullbuttons		use full buttons\n"
	       "\t--nocancel			no cancel button\n"
	       "\t--yes-button <text>		set text of yes button\n"
	       "\t--no-button <text>		set text of no button\n"
	       "\t--ok-button <text>		set text of ok button\n"
	       "\t--cancel-button <text>		set text of cancel button\n"
	       "\t--noitem			don't display items\n"
	       "\t--notags			don't display tags\n"
	       "\t--separate-output		output one line at a time\n"
	       "\t--output-fd <fd>		output to fd, not stdout\n"
	       "\t--title <title>			display title\n"
	       "\t--backtitle <backtitle>		display backtitle\n"
	       "\t--scrolltext			force vertical scrollbars\n"
	       "\t--topleft			put window in top-left corner\n"
	       "\t-h, --help			print this message\n"
	       "\t-v, --version			print version information\n\n"));
    exit(err ? DLG_ERROR : 0 );
}

static void print_version(void) {
    fprintf (stdout, _("whiptail (newt): %s\n"), VERSION);
}
	     
#if 0
/* FIXME Copied from newt.c
 * Place somewhere better -- dialogboxes? -- amck
 */
int wstrlen(const char *str, int len) {
       mbstate_t ps;
       wchar_t tmp;
       int nchars = 0;

       if (!str) return 0;
      if (!len) return 0;
       if (len < 0) len = strlen(str);
      memset(&ps,0,sizeof(mbstate_t));
       while (len > 0) {
               int x,y;

               x = mbrtowc(&tmp,str,len,&ps);
               if (x >0) {
                       str += x;
                       len -= x;
                       y = wcwidth(tmp);
                       if (y>0)
                         nchars+=y;
               } else break;
       }
       return nchars;
}
#endif

/*
 * The value of *width is increased if it is not as large as the width of
 * the line.
 */
static const char * lineWidth(int * width, const char * line, int *chrs)
{
    const char *    s = line;

    if ( line == NULL )
       return 0;

   while ( *s != '\0' && *s != '\n' )
       s++;

    if ( *s == '\n' )
       s++;

    *chrs = _newt_wstrlen (line, s - line );
    *width = max(*width, *chrs);

    return s;
}


/*
 * cleanNewlines
 * Handle newlines in text. Hack.
 */
void cleanNewlines (char *text)
{
    char *p, *q;

    for (p = q = text; *p; p++, q++)
        if (*p == '\\' && p[1] == 'n') {
            p++;
            *q = '\n';
        } else
            *q = *p;
    *q = '\0';
}

/*
 * The height of a text string is added to height, and width is increased
 * if it is not big enough to store the text string.
 */
static const char * textSize(int * height, int * width,
                            int maxWidth,
                            const char * text)
{
    int h = 0;
    int w = 0;
    int chrs = 0;


    if ( text == NULL )
       return 0;

   while ( *text != '\0' ) {
       h++;
       text = lineWidth(width, text, &chrs);
       /* Allow for text overflowing. May overestimate a bit */
       h += chrs /  maxWidth;
    }

    h += 2;
   w += 2;

    *height += h;
    *width += w;

    *width = min(*width, maxWidth);
    return text;
}

/*
 * Add space for buttons.
 * NOTE: when this is internationalized, the button width might change.
 */
static void spaceForButtons(int * height, int * width, int count, int full) {
    /* Make space for the buttons */
    if ( full ) {
       *height += 4;
       if ( count == 1 )
           *width = max(*width, 7);
       else
           *width = max(*width, 20);
    }
    else {
       *height += 2;
       if ( count == 1 )
           *width = max(*width, 7);
       else
           *width = max(*width, 19);
    }
}

static int menuSize(int * height, int * width, int * listHeight,
                    enum mode mode, int * flags, poptContext options) {
    const char ** argv = poptGetArgs(options);
    int         h = 0;
    int         tagWidth = 0;
    int         descriptionWidth = 0;
    int         overhead = 10;

    if ( argv == 0 || *argv == 0 )
       return 0;

    if ( mode == MODE_MENU )
       overhead = 5;

    while ( argv[0] != 0 && argv[1] ) {
       tagWidth = max(tagWidth, _newt_wstrlen(argv[0], -1));
       descriptionWidth = max(descriptionWidth, _newt_wstrlen(argv[1], -1));

       if (mode == MODE_MENU || *flags & FLAG_NOITEM)
           argv += 2;
       else
          argv += 3;
       h++;
    }

    if (*flags & FLAG_NOTAGS) {
	tagWidth = 0;
	overhead -= mode == MODE_MENU ? 1 : 2;
    }
    if (*flags & FLAG_NOITEM) {
	descriptionWidth = 0;
    }

    *width = max(*width, tagWidth + descriptionWidth + overhead);
    *width = min(*width, SLtt_Screen_Cols);

    h = min(h, SLtt_Screen_Rows - *height - 4);
    *height = *height + h + 1;
    *listHeight = h;
    return 0;
}

/*
 * Guess the size of a window, given what will be displayed within it.
 */
static void guessSize(int * height, int * width, int * listHeight,
                     enum mode mode, int * flags, int fullButtons,
                     const char * title, const char * text,
                     poptContext options) {

    int w = 0, h = 0, chrs = 0;

    textSize(&h, &w, SLtt_Screen_Cols -4 , text);     /* Width and height for text */
    lineWidth(&w, title, &chrs);             /* Width for title */

    if ( w > 0 )
       w += 4;

    switch ( mode ) {
       case MODE_CHECKLIST:
       case MODE_RADIOLIST:
       case MODE_MENU:
           spaceForButtons(&h, &w, *flags & FLAG_NOCANCEL ? 1 : 2,
            fullButtons);
           menuSize(&h, &w, listHeight, mode, flags, options);
               break;
       case MODE_YESNO:
           spaceForButtons(&h, &w, 2, fullButtons);
           break;
       case MODE_MSGBOX:
           spaceForButtons(&h, &w, 1, fullButtons);
           break;
       case MODE_INPUTBOX:
           spaceForButtons(&h, &w, *flags & FLAG_NOCANCEL ? 1 : 2,
            fullButtons);
           h += 1;
           break;
       case MODE_GAUGE:
           h += 2;
           break;
       case MODE_NONE:
	   break;
       default:
               break;
    };

    /*
     * Fixed window-border overhead.
     * NOTE: This will change if we add a way to turn off drop-shadow and/or
     * box borders. That would be desirable for display-sized screens.
     */
    w += 2;
    h += 2;

   if ( h > SLtt_Screen_Rows - 1 ) {
       h = SLtt_Screen_Rows - 1;
       *flags |= FLAG_SCROLL_TEXT;
       w += 2; /* Add width of slider - is this right? */
    }

    *width = min(max(*width, w), SLtt_Screen_Cols);
    *height = max(*height, h);
}

char *
readTextFile(const char * filename)
{
    int fd = open(filename, O_RDONLY, 0);
    struct stat s;
    char * buf;

    if ( fd < 0 || fstat(fd, &s) != 0 ) {
       perror(filename);
       exit(DLG_ERROR);
     }

    if ( (buf = malloc(s.st_size + 1)) == 0 ) {
       fprintf(stderr, _("%s: too large to display.\n"), filename);
       exit(DLG_ERROR);
    }

    if ( read(fd, buf, s.st_size) != s.st_size ) {
        perror(filename);
        exit(DLG_ERROR);
    }
   close(fd);
   buf[s.st_size] = '\0';
   return buf;
}

int main(int argc, const char ** argv) {
    enum mode mode = MODE_NONE;
    poptContext optCon;
    int arg;
    char * text;
    const char * nextArg;
    char * end;
    int height;
    int width;
    int listHeight = 1;
    int fd = -1;
    int needSpace = 0;
    int noCancel = 0;
    int noTags = 0;
    int noItem = 0;
    int clear = 0;
    int scrollText = 0;
    int rc = DLG_CANCEL;
    int flags = 0;
    int defaultNo = 0;
    int separateOutput = 0;
    int fullButtons = 0;
    int outputfd = 2;
    int topLeft = 0;
    FILE *output;
    char * result;
    char ** selections, ** next;
    char * title = NULL;
    char *default_item = NULL;
    char * backtitle = NULL;
    char * yes_button = NULL;
    char * no_button = NULL;
    char * ok_button = NULL;
    char * cancel_button = NULL;
    int help = 0, version = 0;
    struct poptOption optionsTable[] = {
	    { "backtitle", '\0', POPT_ARG_STRING, &backtitle, 0 },
	    { "checklist", '\0', 0, 0, OPT_CHECKLIST },
	    { "clear", '\0', 0, &clear, 0 },
	    { "defaultno", '\0', 0, &defaultNo, 0 },
	    { "inputbox", '\0', 0, 0, OPT_INPUTBOX },
	    { "fb", '\0', 0, 0, OPT_FULLBUTTONS },
	    { "fullbuttons", '\0', 0, 0, OPT_FULLBUTTONS },
	    { "gauge", '\0', 0, 0, OPT_GAUGE },
	    { "infobox", '\0', 0, 0, OPT_INFOBOX },
	    { "menu", '\0', 0, 0, OPT_MENU },
	    { "msgbox", '\0', 0, 0, OPT_MSGBOX },
	    { "nocancel", '\0', 0, &noCancel, 0 },
	    { "noitem", '\0', 0, &noItem, 0 },
            { "default-item", '\0', POPT_ARG_STRING, &default_item, 0},
	    { "notags", '\0', 0, &noTags, 0 },
	    { "radiolist", '\0', 0, 0, OPT_RADIOLIST },
	    { "scrolltext", '\0', 0, &scrollText, 0 },
	    { "separate-output", '\0', 0, &separateOutput, 0 },
	    { "title", '\0', POPT_ARG_STRING, &title, 0 },
	    { "textbox", '\0', 0, 0, OPT_TEXTBOX },
	    { "topleft", '\0', 0, &topLeft, 0 },
	    { "yesno", '\0', 0, 0, OPT_YESNO },
	    { "passwordbox", '\0', 0, 0, OPT_PASSWORDBOX },
	    { "output-fd", '\0',  POPT_ARG_INT, &outputfd, 0 },
	    { "yes-button", '\0', POPT_ARG_STRING, &yes_button, 0},
	    { "no-button", '\0', POPT_ARG_STRING, &no_button, 0},
	    { "ok-button", '\0', POPT_ARG_STRING, &ok_button, 0},
	    { "cancel-button", '\0', POPT_ARG_STRING, &cancel_button, 0},
	    { "help", 'h', 0,  &help, 0, NULL, NULL },
	    { "version", 'v', 0, &version, 0, NULL, NULL },
	    { 0, 0, 0, 0, 0 } 
    };
   
#ifdef ENABLE_NLS
    setlocale (LC_ALL, "");
    bindtextdomain (PACKAGE, LOCALEDIR);
    textdomain (PACKAGE);
#endif

    optCon = poptGetContext("whiptail", argc, argv, optionsTable, 0);

    while ((arg = poptGetNextOpt(optCon)) > 0) {
	switch (arg) {
	  case OPT_INFOBOX:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_INFOBOX;
	    break;

	  case OPT_MENU:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_MENU;
	    break;

	  case OPT_MSGBOX:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_MSGBOX;
	    break;

	  case OPT_TEXTBOX:
    	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_TEXTBOX;
	    break;

	  case OPT_PASSWORDBOX:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_PASSWORDBOX;
	    break;

	  case OPT_RADIOLIST:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_RADIOLIST;
	    break;

	  case OPT_CHECKLIST:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_CHECKLIST;
	    break;

	  case OPT_FULLBUTTONS:
	    fullButtons = 1;
	    useFullButtons(1);
	    break;

	  case OPT_YESNO:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_YESNO;
	    break;

	  case OPT_GAUGE:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_GAUGE;
	    break;

	  case OPT_INPUTBOX:
	    if (mode != MODE_NONE) usage(WAS_ERROR);
	    mode = MODE_INPUTBOX;
	    break;
	}
    }
    
    if (help) {
	    usage(NO_ERROR);
	    exit(0);
    }
    if (version) {
	    print_version();
	    exit(0);
    }
    
    if (arg < -1) {
	fprintf(stderr, "%s: %s\n", 
		poptBadOption(optCon, POPT_BADOPTION_NOALIAS), 
		poptStrerror(arg));
	exit(1);
    }

    output = fdopen (outputfd, "w");
    if (output == NULL ) {
        perror ("Cannot open output-fd\n");
        exit (DLG_ERROR);
    }

    if (mode == MODE_NONE) usage(WAS_ERROR);

    if (!(nextArg = poptGetArg(optCon))) usage(WAS_ERROR);
    text = strdup(nextArg);

    if (mode == MODE_TEXTBOX) {
	char *t = text;
	text = readTextFile(t);
	free(t);
    }

    if (!(nextArg = poptGetArg(optCon))) usage(WAS_ERROR);
    height = strtoul(nextArg, &end, 10);
    if (*end) usage(WAS_ERROR);

    if (!(nextArg = poptGetArg(optCon))) usage(WAS_ERROR);
    width = strtoul(nextArg, &end, 10);
    if (*end) usage(WAS_ERROR);

    switch (mode) {
      case MODE_MENU:
      case MODE_RADIOLIST:
      case MODE_CHECKLIST:
	if (!(nextArg = poptGetArg(optCon))) usage(WAS_ERROR);
	listHeight = strtoul(nextArg, &end, 10);
	if (*end) usage(WAS_ERROR);
	break;

      case MODE_GAUGE:
	fd = dup(0);
	if (fd < 0 || close(0) < 0) {
	    perror("dup/close stdin");
	    exit(DLG_ERROR);
	}
	if (open("/dev/tty", O_RDWR) != 0) {
	    perror("open /dev/tty");
	    exit(DLG_ERROR);
	}
	break;
    }

    if (noCancel) flags |= FLAG_NOCANCEL;
    if (noItem) flags |= FLAG_NOITEM;
    if (noTags) flags |= FLAG_NOTAGS;
    if (scrollText) flags |= FLAG_SCROLL_TEXT;
    if (defaultNo) flags |= FLAG_DEFAULT_NO;

    newtInit();
    newtCls();

    cleanNewlines(text);

    if (height <= 0 || width <= 0 || listHeight <= 0) {
	guessSize(&height, &width, &listHeight, mode, &flags, fullButtons,
		 title, text, optCon);
	if (!topLeft && backtitle && SLtt_Screen_Rows > 10 &&
	    height + 1 == SLtt_Screen_Rows &&
	    _newt_wstrlen(backtitle, -1) > (SLtt_Screen_Cols - width) / 2) {
	    height -= 1;
	    listHeight -= 1;
	}
    }

    width -= 2;
    height -= 2;

    newtOpenWindow(topLeft ? 1 : (SLtt_Screen_Cols - width) / 2,
		   topLeft ? 1 : (SLtt_Screen_Rows - height) / 2, width, height, title);
    if (backtitle)
	newtDrawRootText(0, 0, backtitle);

    if (ok_button)
	setButtonText(ok_button, BUTTON_OK);
    if (cancel_button)
	setButtonText(cancel_button, BUTTON_CANCEL);
    if (yes_button)
	setButtonText(yes_button, BUTTON_YES);
    if (no_button)
	setButtonText(no_button, BUTTON_NO);

    switch (mode) {
      case MODE_MSGBOX:
      case MODE_TEXTBOX:
	rc = messageBox(text, height, width, MSGBOX_MSG, flags);
	break;

      case MODE_INFOBOX:
	rc = messageBox(text, height, width, MSGBOX_INFO, flags);
	break;

      case MODE_YESNO:
	rc = messageBox(text, height, width, MSGBOX_YESNO, flags);
	break;

      case MODE_INPUTBOX:
	rc = inputBox(text, height, width, optCon, flags, &result);
	if (rc == DLG_OKAY) {
	    fprintf(output, "%s", result);
	    free(result);
	}
	break;

      case MODE_PASSWORDBOX:
	rc = inputBox(text, height, width, optCon, flags | FLAG_PASSWORD,
	      &result);
	if (rc == DLG_OKAY) {
	    fprintf (output, "%s", result);
	    free(result);
	}
	break;

      case MODE_MENU:
	rc = listBox(text, height, width, listHeight, optCon, flags, default_item, &result);
	if (rc == DLG_OKAY) {
	    fprintf(output, "%s", result);
	    free(result);
	}
	break;

      case MODE_RADIOLIST:
	rc = checkList(text, height, width, listHeight, optCon, 1, flags, &selections);
	if (rc == DLG_OKAY && selections[0]) {
	    fprintf(output, "%s", selections[0]);
	    free(selections[0]);
	    free(selections);
	}
	break;

      case MODE_CHECKLIST:
	rc = checkList(text, height, width, listHeight, optCon, 0, flags, &selections);

	if (!rc) {
	    for (next = selections; *next; next++) {
		if (!separateOutput) {
		    if (needSpace) putc(' ', output);
		    fprintf(output, "\"%s\"", *next);
		    needSpace = 1;
		} else {
		    fprintf(output, "%s\n", *next);
		}
		free(*next);
	    }

	    free(selections);
	}
	break;

      case MODE_GAUGE:
	rc = gauge(text, height, width, optCon, fd, flags);
	break;

      default:
	usage(WAS_ERROR);
    }

    if (rc == DLG_ERROR) usage(WAS_ERROR);

    if (clear)
	newtPopWindow();
    newtFinished();

    free(text);
    poptFreeContext(optCon);

    return ( rc == DLG_ESCAPE ) ? -1 : rc;
}
